File: wfm-develop mobile 07.05.2026/wfm/mobile/android/app/src/main/java/com/beyondviolet/wfm/features/settings/SettingsViewModel.kt
package com.beyondviolet.wfm.features.settings

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.beyondviolet.wfm.core.analytics.AnalyticsEvent
import com.beyondviolet.wfm.core.analytics.AnalyticsService
import com.beyondviolet.wfm.core.managers.ImpersonationStorage
import com.beyondviolet.wfm.core.managers.UserManager
import com.beyondviolet.wfm.core.network.UserService
import com.beyondviolet.wfm.core.network.ApiResponse
import com.beyondviolet.wfm.core.models.Assignment
import com.beyondviolet.wfm.core.models.CurrentShift
import com.beyondviolet.wfm.core.models.ShiftStatus
import com.beyondviolet.wfm.core.models.UserMe
import com.beyondviolet.wfm.feature.auth.data.local.TokenStorage
import com.beyondviolet.wfm.ui.components.ToastManager
import com.beyondviolet.wfm.ui.components.WfmToastState
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.SharingStarted
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.stateIn
import kotlinx.coroutines.launch

/**
 * ViewModel для экрана настроек (профиля пользователя)
 *
 * Управляет состоянием профиля:
 * - Данные текущего пользователя (ФИО, фото, должность)
 * - Загрузка данных
 * - Обновление данных (Pull-to-Refresh)
 * - Выход из аккаунта (logout)
 * - Переключение между назначениями
 * - Режим «Войти как» (impersonation, только для разработчиков с flags.dev в JWT)
 */
open class SettingsViewModel(
    private val userManager: UserManager,
    private val toastManager: ToastManager,
    private val tokenStorage: TokenStorage,
    private val impersonationStorage: ImpersonationStorage,
    private val analyticsService: AnalyticsService,
    private val userService: UserService
) : ViewModel() {

    /**
     * Текущий пользователь
     */
    open val currentUser: StateFlow<UserMe?> = userManager.currentUser
        .stateIn(
            scope = viewModelScope,
            started = SharingStarted.WhileSubscribed(5000),
            initialValue = null
        )

    /**
     * Текущий выбранный assignment
     */
    open val currentAssignment: StateFlow<Assignment?> = userManager.currentAssignment
        .stateIn(
            scope = viewModelScope,
            started = SharingStarted.WhileSubscribed(5000),
            initialValue = null
        )

    /**
     * Текущая смена
     */
    open val currentShift: StateFlow<CurrentShift?> = userManager.currentShift
        .stateIn(
            scope = viewModelScope,
            started = SharingStarted.WhileSubscribed(5000),
            initialValue = null
        )

    /**
     * Индикатор загрузки
     */
    open val isLoading: StateFlow<Boolean> = userManager.isLoading
        .stateIn(
            scope = viewModelScope,
            started = SharingStarted.WhileSubscribed(5000),
            initialValue = false
        )

    /**
     * Индикатор обновления (Pull-to-Refresh)
     */
    open val isRefreshing: StateFlow<Boolean> = userManager.isLoading
        .stateIn(
            scope = viewModelScope,
            started = SharingStarted.WhileSubscribed(5000),
            initialValue = false
        )

    /**
     * Есть ли у текущего пользователя флаг flags.dev в JWT
     */
    private val _isDevUser = MutableStateFlow(false)
    val isDevUser: StateFlow<Boolean> = _isDevUser.asStateFlow()

    /**
     * Активный номер телефона для impersonation (null — режим выключен)
     */
    val impersonationPhone: StateFlow<String?> = impersonationStorage.phone

    /**
     * Индикатор загрузки при удалении аккаунта
     */
    private val _isDeleting = MutableStateFlow(false)
    open val isDeleting: StateFlow<Boolean> = _isDeleting.asStateFlow()

    init {
        viewModelScope.launch {
            val token = tokenStorage.getAccessToken()
            if (token != null) {
                _isDevUser.value = impersonationStorage.isDevUser(token)
            }
        }
    }

    /**
     * Задать номер телефона для impersonation («Войти как»).
     * Пустая строка или null — сбрасывает режим impersonation.
     */
    fun setImpersonationPhone(phone: String?) {
        impersonationStorage.setPhone(phone)
    }

    /**
     * Показывать ли кнопку "Назначения"
     */
    fun shouldShowAssignmentsButton(): Boolean {
        val user = currentUser.value ?: return false
        return user.assignments.size > 1
    }

    /**
     * Можно ли переключаться между назначениями (смена должна быть не открыта)
     */
    fun canSwitchAssignments(): Boolean {
        return currentShift.value?.status != ShiftStatus.OPENED
    }

    /**
     * Проверить возможность переключения назначений
     * Показывает тост если смена открыта
     */
    fun checkCanSwitchAssignments(): Boolean {
        if (!canSwitchAssignments()) {
            toastManager.show("Сначала закройте текущую смену.", state = WfmToastState.ERROR)
            return false
        }
        return true
    }

    /**
     * Переключить текущий assignment
     *
     * Сохраняет выбранный assignment ID и перезагружает данные
     */
    suspend fun switchAssignment(assignmentId: Int) {
        userManager.switchAssignment(assignmentId)
    }

    /**
     * Обновить данные пользователя (Pull-to-Refresh)
     */
    open fun refresh() {
        viewModelScope.launch {
            userManager.refresh()
            val error = userManager.error.value
            if (error != null) {
                toastManager.show(error, state = WfmToastState.ERROR)
            }
        }
    }

    /**
     * Вызывается при появлении экрана
     *
     * Трекает аналитику и обновляет данные пользователя
     */
    fun onAppear() {
        analyticsService.track(AnalyticsEvent.SettingsViewed)
        refresh()
    }

    /**
     * Удалить учётную запись текущего пользователя
     *
     * Вызывает svc_users DELETE /users/me → Beyond Violet Shopping API.
     * onSuccess вызывается после успешного удаления — навигация на экран авторизации.
     * onError вызывается с текстом ошибки при неудаче.
     */
    open fun deleteAccount(onSuccess: () -> Unit, onError: (String) -> Unit) {
        viewModelScope.launch {
            _isDeleting.value = true
            val result = userService.deleteAccount()
            _isDeleting.value = false
            when (result) {
                is ApiResponse.Success -> {
                    viewModelScope.launch { userManager.clear() }
                    onSuccess()
                }
                is ApiResponse.Error -> {
                    onError(result.message ?: "Не удалось удалить аккаунт. Попробуйте позже или обратитесь в поддержку.")
                }
            }
        }
    }

    /**
     * Выход из аккаунта
     *
     * Очищает данные пользователя.
     * Навигация к экрану авторизации выполняется в UI.
     */
    open fun logout() {
        analyticsService.track(AnalyticsEvent.LogoutTapped)
        analyticsService.resetUser()
        impersonationStorage.setPhone(null) // Очищаем режим "Войти как"
        viewModelScope.launch {
            userManager.clear()
        }
    }
}

File: wfm-develop mobile 07.05.2026/wfm/mobile/ios/WFMApp/Features/Settings/SettingsViewModel.swift
import Foundation
import WFMAuth
import WFMUI

/// ViewModel для экрана настроек (профиля пользователя)
///
/// Управляет состоянием профиля:
/// - Данные текущего пользователя (ФИО, фото, должность)
/// - Загрузка данных
/// - Обновление данных (Pull-to-Refresh)
/// - Выход из аккаунта (logout)
/// - Переключение между назначениями
/// - Режим «Войти как» (impersonation, только для разработчиков с flags.dev в JWT)
@MainActor
final class SettingsViewModel: ObservableObject {
    // MARK: - Published State

    /// Текущий пользователь
    @Published private(set) var currentUser: UserMe?

    /// Текущий выбранный assignment
    @Published private(set) var currentAssignment: Assignment?

    /// Текущая смена
    @Published private(set) var currentShift: CurrentShift?

    /// Есть ли у текущего пользователя флаг flags.dev в JWT
    @Published private(set) var isDevUser: Bool = false

    /// Активный номер телефона для impersonation (nil — режим выключен)
    @Published private(set) var impersonationPhone: String?

    /// Флаг загрузки данных (для предотвращения дублирующих запросов)
    @Published private var isRefreshing = false

    // MARK: - Dependencies

    private let userManager: UserManager
    private let toastManager: ToastManager
    private let tokenStorage: TokenStorage
    private let impersonationStorage: ImpersonationStorage
    private let analyticsService: AnalyticsService
    private let userService: UserService

    // Task для отмены предыдущих запросов
    private var loadUserTask: _Concurrency.Task<Void, Never>?
    // MARK: - Initialization

    init(
        userManager: UserManager,
        toastManager: ToastManager,
        tokenStorage: TokenStorage,
        impersonationStorage: ImpersonationStorage,
        analyticsService: AnalyticsService,
        userService: UserService
    ) {
        self.userManager = userManager
        self.toastManager = toastManager
        self.tokenStorage = tokenStorage
        self.impersonationStorage = impersonationStorage
        self.analyticsService = analyticsService
        self.userService = userService
        self.impersonationPhone = impersonationStorage.phone

        // Подписываемся на изменения пользователя, assignment и смены
        _Concurrency.Task {
            for await user in userManager.$currentUser.values {
                currentUser = user
            }
        }

        _Concurrency.Task {
            for await assignment in userManager.$currentAssignment.values {
                currentAssignment = assignment
            }
        }

        _Concurrency.Task {
            for await shift in userManager.$currentShift.values {
                currentShift = shift
            }
        }

        // Проверяем флаг dev в текущем токене
        _Concurrency.Task {
            if let token = try? await tokenStorage.getAccessToken() {
                isDevUser = impersonationStorage.isDevUser(accessToken: token)
            }
        }
    }

    // MARK: - Computed Properties

    /// Показывать ли кнопку "Назначения"
    var shouldShowAssignmentsButton: Bool {
        guard let user = currentUser else { return false }
        return user.assignments.count > 1
    }

    /// Можно ли переключаться между назначениями (смена должна быть не открыта)
    var canSwitchAssignments: Bool {
        currentShift?.status != .opened
    }

    // MARK: - Public Methods

    /// Вызывается при появлении экрана
    ///
    /// Использует HTTP кэширование: сначала показывает кэшированные данные, потом обновляет свежими.
    func loadUser() async {
        // Отменяем предыдущую загрузку
        loadUserTask?.cancel()

        isRefreshing = true
        analyticsService.track(.settingsViewed)

        loadUserTask = _Concurrency.Task {
            await userManager.refresh()

            if let error = userManager.error {
                if currentUser == nil {
                    toastManager.show(message: error, state: .error)
                }
            }
        }

        await loadUserTask?.value
        isRefreshing = false
    }

    /// Обновить данные пользователя (Pull-to-Refresh)
    func refresh() async {
        await loadUser()
    }

    /// Переключить текущий assignment
    ///
    /// Сохраняет выбранный assignment ID и перезагружает данные
    func switchAssignment(assignmentId: Int) async {
        await userManager.switchAssignment(assignmentId: assignmentId)
    }

    /// Задать номер телефона для impersonation («Войти как»).
    ///
    /// Пустая строка или nil — сбрасывает режим impersonation.
    func setImpersonationPhone(_ phone: String?) {
        impersonationStorage.phone = phone?.isEmpty == false ? phone : nil
        impersonationPhone = impersonationStorage.phone
    }

    /// Удалить учётную запись текущего пользователя
    ///
    /// Вызывает svc_users DELETE /users/me → Beyond Violet Shopping API.
    /// После успешного вызова навигация к экрану авторизации через AppRouter.logout().
    func deleteAccount() async throws {
        try await userService.deleteAccount()
    }

    /// Выход из аккаунта
    ///
    /// Очищает данные пользователя.
    /// Навигация к экрану авторизации выполняется в UI через AppRouter.
    func logout() async {
        analyticsService.track(.logoutTapped)
        analyticsService.resetUser()
        impersonationStorage.phone = nil // Очищаем режим "Войти как"
        await userManager.clear()
    }
}
